roachtest: add support for collecting Go coverage

RaduBerinde opened the pull request:
PR #110280 added instrumentation to the cockroach binary. When built in this mode, cockroach dumps coverage counters to the directory specified in `BAZEL_COVER_DIR` env var. Note that this is similar to Go's recent support for binary instrumentation
(https://go.dev/testing/coverage/); we will eventually switch to using that when Bazel supports building in that mode.

This commit adds a `--go-cover` flag to `roachtest`. When this flag is set, the `BAZEL_COVER_DIR` env var is set and after the test completes we copy the contents to the artifacts (similar to the `perf` artifacts).

Epic: none
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/110434)
<!-- Reviewable:end -->

RaduBerinde commented:
An iteration of this change was tested manually. I plan to test again from a teamcity build before merging.

renatolabs commented on pkg/cmd/roachtest/test_runner.go:
```diff
@@ -913,6 +917,47 @@ fi'`
+// getGoCoverArtifacts retrieves the go coverage artifacts for the test.
+// If there's an error, oh well, don't do anything rash like fail a test
+// which already passed.
+func getGoCoverArtifacts(ctx context.Context, l *logger.Logger, c *clusterImpl, t test.Test) {
```
Unless I'm missing something, sharing a function for fetching perf and coverage artifacts would be very simple so I don't see why not.

renatolabs commented on pkg/cmd/roachtest/test_runner.go:
```diff
@@ -913,6 +917,47 @@ fi'`
+else
+    echo false
+fi'`
+			result, err := c.RunWithDetailsSingleNode(ctx, t.L(), c.Node(node), "bash", "-c", testCmd)
```
Nit: `bash -c` shouldn't be needed. `RunWithDetailsSingleNode` already adds `bash -c`.

renatolabs commented on pkg/cmd/roachtest/test_runner.go:
```diff
@@ -1273,24 +1318,48 @@ func (r *testRunner) teardownTest(
+		// soon as possible. We send two SIGINTs; the first one starts draining and
+		// the second causes the process to exit right away.
+		//
+		// TODO(radu): many tests stop the nodes with the default options
+		// (SIGKILL) during the test, which means some coverage data will be lost.
```
Is there anything we can do in this case? Just wondering if there's anything actionable from this TODO (maybe it should just be a note).

Also: what happens in the case of tests that restart nodes gracefully (SIGINT)? Will the coverage data reflect before and after restart? Or just after?

renatolabs reviewed (commented):
This is great 👍

renatolabs commented:
Are we going to update the nightly scripts to pass the new flag in follow-up work? Maybe it would be better to do both at the same time so that we could run some test runs on TC (with a subset of tests, using `SELECT_PROBABILITY`).

renatolabs commented on pkg/cmd/roachtest/test_runner.go:
```diff
@@ -1273,24 +1318,48 @@ func (r *testRunner) teardownTest(
+			t.L().PrintfCtx(ctx, "error stopping cluster: %v", err)
+		}
+		t.L().Printf("Retrieving go cover artifacts")
+		getGoCoverArtifacts(ctx, t.L(), c, t)
```
I don't get it. Why are we only fetching coverage data when the test failed for non-timeout reasons? Aren't we interested in coverage data for all runs?

RaduBerinde reviewed (commented):
I will set up a separate build under the Coverage project and I will run it against this change together with some new scripts in a draft PR.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110434#-:-NeAhDAj7nMVkBL7O333:b-cggdop)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @DarrylWong, @rachitgsrivastava, @renatolabs, and @srosenberg)

___
*[`pkg/cmd/roachtest/test_runner.go` line 1347 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/110434#gh-1323630808:-NeAggNE-KMaPSRGWa3G:by0d0mu) ([raw file](https://github.com/cockroachdb/cockroach/blob/b539121ad2f3f27234e82df168cb2ef048667859/pkg/cmd/roachtest/test_runner.go#L1347)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

Is there anything we can do in this case? Just wondering if there's anything actionable from this TODO (maybe it should just be a note).

Also: what happens in the case of tests that restart nodes gracefully (SIGINT)? Will the coverage data reflect before and after restart? Or just after?
</blockquote></details>

Not sure what can be done, that's really what the TODO is for. Perhaps changing the Stop flags so that the default is to stop it with a double SIGINT.

Each run generates a different file (there's a random number in it). So we'll get both the before and after coverage.
___
*[`pkg/cmd/roachtest/test_runner.go` line 1362 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/110434#gh-1323637967:-NeAh1gF8MUks_Zgpvtp:b-xhdavq) ([raw file](https://github.com/cockroachdb/cockroach/blob/b539121ad2f3f27234e82df168cb2ef048667859/pkg/cmd/roachtest/test_runner.go#L1362)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

I don't get it. Why are we only fetching coverage data when the test failed for non-timeout reasons? Aren't we interested in coverage data for all runs?
</blockquote></details>

We're only fetching it on success (it's the else to `timedOut || t.Failed()`).


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
*[`pkg/cmd/roachtest/test_runner.go` line 936 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/110434#gh-1323607341:-NeEHu8t3aXY4X3aKpMO:bx0txx0) ([raw file](https://github.com/cockroachdb/cockroach/blob/b539121ad2f3f27234e82df168cb2ef048667859/pkg/cmd/roachtest/test_runner.go#L936)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

Nit: `bash -c` shouldn't be needed. `RunWithDetailsSingleNode` already adds `bash -c`.
</blockquote></details>

Are you sure about this? It seems like it's not added if the cluster is local:
https://github.com/cockroachdb/cockroach/blob/master/pkg/roachprod/install/cluster_synced.go#L970


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110434)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @DarrylWong, @rachitgsrivastava, @renatolabs, and @srosenberg)

___
*[`pkg/cmd/roachtest/test_runner.go` line 923 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/110434#gh-1323607041:-NeEJWXkDqdHj9kQqvxv:b9y4ise) ([raw file](https://github.com/cockroachdb/cockroach/blob/b539121ad2f3f27234e82df168cb2ef048667859/pkg/cmd/roachtest/test_runner.go#L923)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

Unless I'm missing something, sharing a function for fetching perf and coverage artifacts would be very simple so I don't see why not.
</blockquote></details>

Done, I also cleaned up the code a bit.
___
*[`pkg/cmd/roachtest/test_runner.go` line 1362 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/110434#gh-1323637967:-NeEJdlc2gdEaeETS31m:b-4wlzix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b539121ad2f3f27234e82df168cb2ef048667859/pkg/cmd/roachtest/test_runner.go#L1362)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

We're only fetching it on success (it's the else to `timedOut || t.Failed()`).
</blockquote></details>

Refactored this code so it's more clear what's going on.


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
*[`pkg/cmd/roachtest/test_runner.go` line 936 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/110434#gh-1323607341:-NeEKhd_02BTBAYu9Jq5:bcq53bx) ([raw file](https://github.com/cockroachdb/cockroach/blob/b539121ad2f3f27234e82df168cb2ef048667859/pkg/cmd/roachtest/test_runner.go#L936)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Are you sure about this? It seems like it's not added if the cluster is local:
https://github.com/cockroachdb/cockroach/blob/master/pkg/roachprod/install/cluster_synced.go#L970
</blockquote></details>

Ah, `newSession` adds it for the local case, and in the other case we run `ssh` so we don't need `bash -c` anyway. Removed.


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
*[`pkg/cmd/roachtest/test_runner.go` line 936 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/110434#gh-1323607341:-NeIg86KAGSZOiMdGEI-:bdbohmm) ([raw file](https://github.com/cockroachdb/cockroach/blob/b539121ad2f3f27234e82df168cb2ef048667859/pkg/cmd/roachtest/test_runner.go#L936)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Ah, `newSession` adds it for the local case, and in the other case we run `ssh` so we don't need `bash -c` anyway. Removed.
</blockquote></details>

Nope, it doesn't work without `bash -c` (this is on a test running on GCE):
```
  | stderr:bash: line 8: $'ARTIFACTS_DIR="gocover"\nif [[ -d "${ARTIFACTS_DIR}" ]]; then\n    echo true\nelif [[ -e "${ARTIFACTS_DIR}" ]]; then\n    ls -la "${ARTIFACTS_DIR}"\n    exit 1\nelse\n    echo false\nfi': command not found
```


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Unfortunately not even the "double SIGINT" thing runs the exit hooks.. I will add a handler for a new signal which just calls `os.Exit()` (in a separate PR).

renatolabs commented:
> I will add a handler for a new signal which just calls os.Exit() (in a separate PR).

Instead of adding new code, maybe we can just wait for a graceful shutdown (with a reasonable `MaxWait`). Do we get coverage data in that case?

RaduBerinde reviewed (commented):
We should get coverage in that case but I'm afraid that might slow down many tests. Not even sure what happens if we gracefully shut down all nodes in the cluster, can they even perform a graceful shutdown? I also do want to give tests that kill processes a better option. I was thinking of using `SIGUSR1` and just have it do a `log.Fatal` (which calls `os.Exit`) - that should be the closest thing to a `KILL` that still dumps coverage.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110434#-:-NeORbBK-BBJMVV174EW:b-woo2ao)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @DarrylWong, @rachitgsrivastava, @renatolabs, and @srosenberg)



<!-- Sent from Reviewable.io -->

renatolabs commented:
> I'm afraid that might slow down many tests

Would it? I think a  `MaxWait` of 30s is very reasonable and should work for the vast majority of tests while still being pretty minor relative to the duration of an average test (several minutes to a few hours).

> can they even perform a graceful shutdown

Yes, a single SIGINT will drain and shutdown gracefully. That's how upgrade tests work currently. It's publicly documented behaviour: https://www.cockroachlabs.com/docs/stable/node-shutdown#drain-the-node-and-terminate-the-node-process.

Adding custom logic to SIGUSR1, especially a `log.Fatal`, is relase-note worthy change that we should avoid IMO.

RaduBerinde commented:
I meant - if all nodes are shutting down at the same time, will graceful termination ever succeed? I thought it tries to move all leaseholders off the node, but that won't work if all nodes try to do that? Not sure of the behavior here.

Yes, SIGUSR1 is a termination signal.

RaduBerinde commented:
Anyway, I'll give it a shot and see in how many tests the graceful shutdown times out

RaduBerinde commented:
As discussed offline, it turns out that cluster-wide graceful shutdown is not expected in general to complete. Updated the change to use `SIGUSR1` which will work in conjunction with https://github.com/cockroachdb/cockroach/pull/110864

srosenberg reviewed (approved):
Nice work! I am really excited to take this for a spin :)

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110434#-:-NepWjNiDbd6TqfdU9Wv:b-dn46vf)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @DarrylWong, @rachitgsrivastava, and @renatolabs)

___
*[`pkg/cmd/roachtest/test_runner.go` line 1347 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/110434#gh-1323630808:-NepVscJAWj0TH21xXyq:b-3kyq2) ([raw file](https://github.com/cockroachdb/cockroach/blob/b539121ad2f3f27234e82df168cb2ef048667859/pkg/cmd/roachtest/test_runner.go#L1347)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Not sure what can be done, that's really what the TODO is for. Perhaps changing the Stop flags so that the default is to stop it with a double SIGINT.

Each run generates a different file (there's a random number in it). So we'll get both the before and after coverage.
</blockquote></details>

I think we need to revisit the business of shutting down nodes. We're starting the process via `systemd` (see start.sh), but we're stopping it via `kill`. Switching to `systemd` (for stopping) would allow for trapping on `SIGTERM`. We also have `StopCockroachGracefullyOnNode`, which sends `SIGTERM` followed by calling `Stop`; the former is used only by a small subset of the roachtests. 


<!-- Sent from Reviewable.io -->

srosenberg reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110434)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @DarrylWong, @rachitgsrivastava, @RaduBerinde, and @renatolabs)

___
*[`pkg/cmd/roachtest/test_runner.go` line 1342 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/110434#-NepYQrNAVag2psAfHVv:-NepYQrNAVag2psAfHVw:bu092lk) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e2d417eb5a23ee4b7e609755250fcb9d9f0cb75/pkg/cmd/roachtest/test_runner.go#L1342)):*
> ```Go
> 		t.L().Printf("Stopping all nodes to obtain go cover artifacts")
> 		stopOpts := option.DefaultStopOpts()
> 		stopOpts.RoachprodOpts.Sig = 10 // SIGUSR1
> ```

This works for majority of the cases. However, several roachtests invoke `cluster.StopE` during a test run, e.g., `chaos.Runner` is used by `cdc` and `tpcc`. This also means if someone is running a manual test via `roachprod`, they won't see coverage dump if `roachprod stop` is invoked. Methinks we should instead modify `StopE` to use `systemd`.


<!-- Sent from Reviewable.io -->

srosenberg reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110434)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @DarrylWong, @rachitgsrivastava, @RaduBerinde, and @renatolabs)

___
*[`pkg/cmd/roachtest/test_runner.go` line 1362 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/110434#gh-1323637967:-Nep_-rkAFCzx5UDN9Nk:btiea71) ([raw file](https://github.com/cockroachdb/cockroach/blob/b539121ad2f3f27234e82df168cb2ef048667859/pkg/cmd/roachtest/test_runner.go#L1362)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Refactored this code so it's more clear what's going on.
</blockquote></details>

Shouldn't we fetch coverage in _all_ cases? What if a test fails in an interesting way, uncovering new execution paths. We would never see those paths covered in successful runs.


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
TFTR!

I have a build going with the latest changes: https://teamcity.cockroachdb.com/viewLog.html?buildId=11854607&buildTypeId=Cockroach_Coverage_RoachtestNightlyGce
Artifacts for each test contain gocover stuff.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110434#-:-Nepp84D36p5PjsFjOOs:b-a6b0tc)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @DarrylWong, @rachitgsrivastava, @renatolabs, and @srosenberg)

___
*[`pkg/cmd/roachtest/test_runner.go` line 1347 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/110434#gh-1323630808:-Nepo3cI3tukD0HrWwGX:b9eyk7c) ([raw file](https://github.com/cockroachdb/cockroach/blob/b539121ad2f3f27234e82df168cb2ef048667859/pkg/cmd/roachtest/test_runner.go#L1347)):*
<details><summary><i>Previously, srosenberg (Stan Rosenberg) wrote…</i></summary><blockquote>

I think we need to revisit the business of shutting down nodes. We're starting the process via `systemd` (see start.sh), but we're stopping it via `kill`. Switching to `systemd` (for stopping) would allow for trapping on `SIGTERM`. We also have `StopCockroachGracefullyOnNode`, which sends `SIGTERM` followed by calling `Stop`; the former is used only by a small subset of the roachtests.
</blockquote></details>

The thing is, you can't shutdown gracefully in all cases (i.e. draining won't finish). There need to be other live nodes to take over your leaseholders. Also, some tests must be testing processes getting brutally killed on purpose (to test crash recovery paths).

I agree though that the default should be the "nice" way and tests should opt into more brutal killing.
___
*[`pkg/cmd/roachtest/test_runner.go` line 1362 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/110434#gh-1323637967:-NepnNMaBdqPfZ-S8eVL:b5am769) ([raw file](https://github.com/cockroachdb/cockroach/blob/b539121ad2f3f27234e82df168cb2ef048667859/pkg/cmd/roachtest/test_runner.go#L1362)):*
<details><summary><i>Previously, srosenberg (Stan Rosenberg) wrote…</i></summary><blockquote>

Shouldn't we fetch coverage in _all_ cases? What if a test fails in an interesting way, uncovering new execution paths. We would never see those paths covered in successful runs.
</blockquote></details>

There was some technical difficulty with collecting in failure cases, I can't remember what it was. Anyway, I wouldn't count a failed test as covering something (the goal should always be to fix failing tests, not keep them failing for better coverage).
___
*[`pkg/cmd/roachtest/test_runner.go` line 1342 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/110434#-NepYQrNAVag2psAfHVv:-Nepommz031MPWiZZFqm:b-59jzyp) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e2d417eb5a23ee4b7e609755250fcb9d9f0cb75/pkg/cmd/roachtest/test_runner.go#L1342)):*
<details><summary><i>Previously, srosenberg (Stan Rosenberg) wrote…</i></summary><blockquote>

This works for majority of the cases. However, several roachtests invoke `cluster.StopE` during a test run, e.g., `chaos.Runner` is used by `cdc` and `tpcc`. This also means if someone is running a manual test via `roachprod`, they won't see coverage dump if `roachprod stop` is invoked. Methinks we should instead modify `StopE` to use `systemd`.
</blockquote></details>

I think in the short term it would be good to refactor the roachtest stop options to be a higher level construct instead of test code directly specifying signals, and that way we could easily switch from SIGKILL to SIGUSR1 to get the coverage.


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
First commit is now #111151

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110434#-:-Nf2wh0X39OB9UF_RH43:bfhl1v)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @DarrylWong, @rachitgsrivastava, @renatolabs, and @srosenberg)



<!-- Sent from Reviewable.io -->

RaduBerinde committed:
roachtest: add support for collecting Go coverage

PR #110280 added instrumentation to the cockroach binary. When built
in this mode, cockroach dumps coverage counters to the directory
specified in `BAZEL_COVER_DIR` env var. Note that this is similar to
Go's recent support for binary instrumentation
(https://go.dev/testing/coverage/); we will eventually switch to using
that when Bazel supports building in that mode.

This commit adds a `-go-cover` flag to `roachtest`. When this flag is
set, the `BAZEL_COVER_DIR` env var is set and after the test completes
we copy the contents to the artifacts (similar to the `perf`
artifacts).

Epic: none
Release note: None

RaduBerinde commented:
Updated to zip the gocoverage files into a separate archive.

RaduBerinde commented:
Sample run (with some script changes): https://teamcity.cockroachdb.com/viewLog.html?buildTypeId=Cockroach_Coverage_RoachtestNightlyGce&buildId=11930658&branch_Cockroach_Coverage_RoachtestNightlyGce=110920

RaduBerinde commented:
bors r+
